src/hammer-vlsi/technology/asap7/__init__.py

```diff
@@ -124,14 +127,19 @@ def scale_gds_script(self, gds_file: str) -> str:
         Note: Need to escape TCL constructs such as [] and {}!
         """
 
-        return """
-#!/usr/bin/python3
+        return """#!/usr/bin/python3
```

Comment:
Why leave this script here, I thought `post_par_script` would work for current case?

Reply:
No, the issue that @colinschmidt raised in the review was that `post_par_script` always ran after the `par` action. However, this is wrong because if you say only ran a subset of `par` steps that did not include writing the output GDS, you would run the script when it should not have been run. So, there is no more `post_par_script`.

The other thing this does is demonstrate how you can use a hook provided by a technology. See ucb-bar/chipyard#263.

Reply:
I realized that this script is still insufficient for hierarchical P&R. It would be wrong to scale down submodule GDSs and then scale them down again at the top-level. Just checking against the modules listed in the hierarchical placement constraints will also be insufficient because deduplication of referenced cells in the GDS will cause those cells to be shrunk again. This may need to be punted and we'll note for now that ASAP7 cannot be used with hierarchical flow.

Reply:
I understand it now.
should we change our method from editing par gds to patch `28nm -> 7nm` lef in `post_install_script`, so that we can directly use real 7nm process.
This may eventually solve our problem.

Reply:
I don't think so, because it will mess up the parasitic extraction if P&R is done at 7nm, unless we also patch the QRC techfile (but I don't know how).

Also, the reason for 28nm was academic Cadence licensing, so I think we should keep it at 28nm so that other schools can still use Hammer with ASAP7.

Reply:
I understand, you are right.
QRC cannot be patched since it is a close-source spec. 
Academic issue is surly the biggest problem.

> this script is still insufficient for hierarchical P&R

Yes that's the problem I used to have, so I set the `gds_merge` default to true.
and also different PaR tool may have different strategy to deliver the gds patterning.
I think we should figure out what should be scaled and what shouldn't to satisfy `gds_merge` both `true` or `false`.

Reply:
Okay, let's think this out. Let's say module `A` is a parent of module `B`. Module `A` contains a blackbox `A_bb` and module `B` contains a blackbox `B_bb`.

After P&R-ing module `B`, all shapes in `B` and `B_bb` should be scaled down.

After P&R-ing module `A`, all shapes in `A` and `A_bb` should be scaled but `B` and `B_bb` should not be scaled.

Perhaps we can write out the list of scaled cells when scaling `B` (in format `cellname_filename` as will be generated by the `-uniquify` option in Innovus), then passing this list to its parent module `A` and appending to the `cell_list` variable in the existing script. We will have to traverse the hierarchical tree to make sure we pull the right files, but at least `gds_merge` can always be `true` or `false`.

Any other ideas?

Reply:
> After P&R-ing module B, all shapes in B and B_bb should be scaled down.

How about making scale down as the final step?
We can just use 28nm during all PaR, and leave the scale problem until tape-in.

Reply:
We want the ability to do hierarchical DRC & LVS, which requires the leaf modules be already scaled down.

Reply:
Emmm, I see. If this is the only reason, can we do like this?
Module `B` netlist → PaR → gds → scale down → DRC & LVS
------------------------------------------↓
--------------------------------------merge → gds → scale down → DRC & LVS
------------------------------------------↑
Module `A` netlist → PaR → gds → scale down → DRC & LVS
But this method might introduce some boundary problem to DRC & LVS after merging.

Reply:
Maybe, but if module `A` and `B` don't have merging done, DRC & LVS won't be valid. We would need to have another merge & scale step (which would create 2 GDSs).

Also, if you scale down `A` and `B` without merging hard macros like analog IP blocks, the final scale down step will need to scale down those hard macros without also scaling down geometries generated in `A` and `B`. I think this is just a variation of the same problem.

Reply:
Yes. This method is only for module level check DRC & LVS,
my approach will add additional trouble to this issue.
I'm read the doc from `INNOVUS/doc/innovusUG/Importing_and_Exporting_Designs.html`
will this give a help for this problem?

Reply:
I think it just supports my thought that we can take advantage of the `-uniquify` appending the module name in front of the cell name to filter out cells that have already been scaled. Other than that, I don't see much else that's helpful in that guide.

Reply:
Yes, I do think so.

Reply:
Let's merge this PR and address #533 in a different PR.